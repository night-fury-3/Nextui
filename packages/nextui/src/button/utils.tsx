import React from 'react';
import { ButtonProps } from '../button/button';
import { ButtonGroupConfig } from '../button/button-group-context';

export const filterPropsWithGroup = (
  props: React.PropsWithChildren<ButtonProps>,
  config: ButtonGroupConfig
): ButtonProps => {
  if (!config.isButtonGroup) return props;
  return {
    ...props,
    auto: true,
    shadow: false,
    bordered: config.bordered || props.bordered,
    ghost: config.ghost || props.ghost,
    flat: config.flat || props.flat,
    animated: config.animated || props.animated,
    rounded: config.rounded || props.rounded,
    light: config.light || props.light,
    size: config.size || props.size,
    color: config.color || props.color,
    disabled: config.disabled || props.disabled,
  };
};
