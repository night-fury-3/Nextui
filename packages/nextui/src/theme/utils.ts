import { NextUIThemes } from './index';
import useWarning from '@hooks/use-warning';
import { DeepPartial } from '@utils/types';
import { deepMergeObject } from '@utils/object';
import darkTheme from './dark';
import lightTheme from './default';

export type PartialTheme = DeepPartial<NextUIThemes>;

export type ThemeParam =
  | PartialTheme
  | ((theme: PartialTheme) => PartialTheme)
  | undefined;

export const mergeTheme = (
  current: NextUIThemes,
  custom: ThemeParam
): NextUIThemes => {
  if (!custom) return current;
  if (typeof custom === 'function') {
    const merged = custom(current);
    if (!merged || typeof merged !== 'object') {
      useWarning('The theme function must return object value.');
    }
    return merged as NextUIThemes;
  }
  return deepMergeObject<NextUIThemes>(current, custom as NextUIThemes);
};

export const switchTheme = (mergedTheme: NextUIThemes): NextUIThemes => {
  const themes: { [key in NextUIThemes['type']]: NextUIThemes } = {
    light: lightTheme,
    dark: darkTheme,
  };
  return { ...mergedTheme, ...themes[mergedTheme.type] };
};
