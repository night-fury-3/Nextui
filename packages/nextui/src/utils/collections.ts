import React, { ReactNode } from 'react';

export const getId = () => {
  return Math.random().toString(32).slice(2, 10);
};

export const hasChild = (
  children: ReactNode | undefined,
  child: React.ElementType
): boolean => {
  const types = React.Children.map(children, (item) => {
    if (!React.isValidElement(item)) return null;
    return item.type;
  });

  return (types || []).includes(child);
};

export const pick = <Obj extends { [key: string]: any }, Key extends keyof Obj>(
  props: Key[],
  obj: Obj
): Pick<Obj, Key> =>
  props.reduce((acc, prop) => {
    acc[prop] = obj[prop];
    return acc;
  }, {} as Pick<Obj, Key>);

export const pickChild = (
  children: ReactNode | undefined,
  targetChild: React.ElementType
): [ReactNode | undefined, ReactNode | undefined] => {
  let target: ReactNode[] = [];
  const withoutTargetChildren = React.Children.map(children, (item) => {
    if (!React.isValidElement(item)) return item;
    if (item.type === targetChild) {
      target.push(item);
      return null;
    }
    return item;
  });

  const targetChildren = target.length >= 0 ? target : undefined;

  return [withoutTargetChildren, targetChildren];
};
