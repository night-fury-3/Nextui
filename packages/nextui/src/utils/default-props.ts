import { NormalSizes } from './prop-types';
import type { CSSProperties } from 'react';

export type NextUISpacingWithAuto = NormalSizes | 'px' | 'auto' | number;

export interface NextUIMargins {
  m?: NextUISpacingWithAuto;
  my?: NextUISpacingWithAuto;
  mx?: NextUISpacingWithAuto;
  mt?: NextUISpacingWithAuto;
  mb?: NextUISpacingWithAuto;
  ml?: NextUISpacingWithAuto;
  mr?: NextUISpacingWithAuto;
}

export interface NextUIPaddings {
  p?: NextUISpacingWithAuto;
  py?: NextUISpacingWithAuto;
  px?: NextUISpacingWithAuto;
  pt?: NextUISpacingWithAuto;
  pb?: NextUISpacingWithAuto;
  pl?: NextUISpacingWithAuto;
  pr?: NextUISpacingWithAuto;
}

export type NextUISpaces = NextUIMargins & NextUIPaddings;

export type NextUISpacesKeys = { [key in keyof NextUISpaces]?: string };

export const NextUISpacesStringKeys = [
  'm',
  'my',
  'mx',
  'mt',
  'mb',
  'ml',
  'mr',
  'p',
  'py',
  'px',
  'pt',
  'pb',
  'pl',
  'pr'
];

export interface DefaultProps extends NextUIMargins, NextUIPaddings {
  style?: CSSProperties;
  className?: string;
}
