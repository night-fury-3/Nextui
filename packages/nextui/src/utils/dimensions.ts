import { NormalWeights, NormalSizes } from './prop-types';

interface GetSizeValue {
  size: string | number;
  sizes: Record<string, any>;
  defaultSize?: string;
}

export const getMargin = (num: number): string => {
  return `calc(${num * 15.25}pt + 1px * ${num - 1})`;
};

export const getNormalWeight = (weight?: NormalWeights): string | undefined => {
  const weights: { [key in NormalWeights]?: string } = {
    light: '1px',
    normal: '2px',
    bold: '3px'
  };
  return weights[weight || 'normal'];
};

export const getNormalRadius = (
  size: NormalSizes,
  rounded?: boolean
): string => {
  const radius: { [key in NormalSizes]: string } = {
    mini: '7px',
    small: '9px',
    medium: '12px',
    large: '14px',
    xlarge: '15px'
  };
  const baseRadius = radius[size];
  return rounded ? `calc(${baseRadius} + 10rem)` : baseRadius;
};

export function getSizeValue({
  size,
  sizes,
  defaultSize = 'md'
}: GetSizeValue) {
  if (typeof size === 'number') {
    return size;
  }
  return sizes[size] || size || sizes[defaultSize];
}
