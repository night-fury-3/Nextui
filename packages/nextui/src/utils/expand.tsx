import React, { useEffect, useRef, useState } from 'react';
import withDefaults from '../utils/with-defaults';
import useRealShape from '../use-real-shape';
import clsx from './clsx';

interface Props {
  isExpanded?: boolean;
  animated?: boolean;
  delay?: number;
}

const defaultProps = {
  isExpanded: false,
  animated: true,
  delay: 200
};

export type ExpandProps = Props & typeof defaultProps;

const Expand: React.FC<React.PropsWithChildren<ExpandProps>> = ({
  isExpanded,
  delay,
  animated,
  children
}) => {
  const [height, setHeight] = useState<string>(isExpanded ? 'auto' : '0');
  const [selfExpanded, setSelfExpanded] = useState<boolean>(isExpanded);
  const contentRef = useRef<HTMLDivElement>(null);
  const entryTimer = useRef<number>();
  const leaveTimer = useRef<number>();
  const resetTimer = useRef<number>();
  const [state, updateShape] = useRealShape<HTMLDivElement>(contentRef);

  useEffect(() => setHeight(`${state.height}px`), [state.height]);
  useEffect(() => {
    // show element or reset height.
    // force an update once manually, even if the element does not change.
    // (the height of the element might be "auto")
    if (!isExpanded) {
      updateShape();
      setHeight(`${state.height}px`);
    }

    // show expand animation
    entryTimer.current = window.setTimeout(() => {
      setSelfExpanded(isExpanded);
      clearTimeout(entryTimer.current);
    }, 30);

    // Reset height after animation
    if (isExpanded) {
      resetTimer.current = window.setTimeout(() => {
        setHeight('auto');
        clearTimeout(resetTimer.current);
      }, delay);
    } else {
      leaveTimer.current = window.setTimeout(() => {
        clearTimeout(leaveTimer.current);
      }, delay / 2);
    }

    return () => {
      clearTimeout(entryTimer.current);
      clearTimeout(leaveTimer.current);
      clearTimeout(resetTimer.current);
    };
  }, [isExpanded]);

  return (
    <div className={clsx('container', { expanded: selfExpanded, animated })}>
      <div ref={contentRef} className="content">
        {children}
      </div>
      <style jsx>{`
        .container {
          padding: 0;
          margin: 0;
          height: 0;
          opacity: 0;
          overflow: hidden;
        }
        .animated {
          transition: height ${delay}ms ease 0ms,
            opacity ${delay * 1.5}ms ease 0ms;
        }
        .expanded {
          height: ${height};
          opacity: 1;
        }
      `}</style>
    </div>
  );
};

export default withDefaults(Expand, defaultProps);
