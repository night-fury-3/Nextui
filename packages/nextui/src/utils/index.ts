import { NormalColors, normalColors } from './prop-types';
import { NextUIThemesPalette } from '@theme/index';
/**
 * This function allows validate if a string is a hexadecimal
 * value
 * @param str [string] hexadecimal value
 * @returns result [boolean]
 */
export const isHex = (str: string): boolean => {
  const exp = /#[a-fA-F0-9]{3,6}/g;
  return exp.test(str);
};

/**
 *
 * @param hex [string]
 * @param alpha [number]
 * @returns [string]
 */
export const hexToRGBA = (hex: string, alpha: number = 1): string => {
  let r: string | number = 0,
    g: string | number = 0,
    b: string | number = 0;
  // 3 digits
  if (hex.length == 4) {
    r = '0x' + hex[1] + hex[1];
    g = '0x' + hex[2] + hex[2];
    b = '0x' + hex[3] + hex[3];
    // 6 digits
  } else if (hex.length == 7) {
    r = '0x' + hex[1] + hex[2];
    g = '0x' + hex[3] + hex[4];
    b = '0x' + hex[5] + hex[6];
  }
  return `rgba(${+r}, ${+g},${+b},${alpha})`;
};

export const getNormalColor = (
  color: NormalColors | string | undefined,
  palette: NextUIThemesPalette,
  defaultColor: string = 'inherit'
) => {
  const colors: { [key in NormalColors | string]: string } = {
    default: defaultColor,
    primary: palette.primary,
    secondary: palette.secondary,
    success: palette.success,
    warning: palette.warning,
    error: palette.error,
    gradient: palette.gradient,
  };
  return color && colors[color] ? colors[color] : defaultColor;
};

export const isNormalColor = (color: string): boolean => {
  let found = normalColors.find((el) => el === color);
  return found !== undefined && found !== null;
};

/**
 * Function that checks color name support in the current browser
 * @param strColor
 * @returns boolean
 */
export const isColor = (strColor: string) => {
  let s = new Option().style;
  s.color = strColor;
  return s.color == strColor;
};
