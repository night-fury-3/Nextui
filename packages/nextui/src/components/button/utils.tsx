import React, { ReactNode } from 'react';
import { NormalSizes } from '@utils/prop-types';
import ButtonIcon from './button-icon';
import { ButtonProps } from '@components/button/button';
import { ButtonGroupConfig } from '@components/button/button-group-context';

export const getButtonChildrenWithIcon = (
  auto: boolean,
  size: NormalSizes,
  children: ReactNode,
  icons: {
    icon?: React.ReactNode;
    iconRight?: React.ReactNode;
  }
) => {
  const { icon, iconRight } = icons;
  const hasIcon = icon || iconRight;
  const isRight = Boolean(iconRight);
  const paddingForAutoMode =
    auto || size === 'mini'
      ? `calc(var(--next-ui-button-height) / 2 + var(--next-ui-button-padding) * .5)`
      : 0;
  if (!hasIcon) return <div className="text">{children}</div>;
  if (React.Children.count(children) === 0) {
    return (
      <ButtonIcon isRight={isRight} isSingle>
        {hasIcon}
      </ButtonIcon>
    );
  }
  return (
    <React.Fragment>
      <ButtonIcon isRight={isRight}>{hasIcon}</ButtonIcon>
      <div className={`text ${isRight ? 'right' : 'left'}`}>
        {children}
        <style jsx>{`
          .left {
            padding-left: ${paddingForAutoMode};
          }
          .right {
            padding-right: ${paddingForAutoMode};
          }
        `}</style>
      </div>
    </React.Fragment>
  );
};

export const filterPropsWithGroup = (
  props: React.PropsWithChildren<ButtonProps>,
  config: ButtonGroupConfig
): ButtonProps => {
  if (!config.isButtonGroup) return props;
  return {
    ...props,
    auto: true,
    shadow: false,
    bordered: config.bordered || props.bordered,
    flat: config.flat || props.flat,
    animated: config.animated || props.animated,
    rounded: config.rounded || props.rounded,
    light: config.light || props.light,
    size: config.size || props.size,
    color: config.color || props.color,
    disabled: config.disabled || props.disabled,
  };
};
