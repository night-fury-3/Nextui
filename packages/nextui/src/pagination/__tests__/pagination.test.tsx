import React from 'react';
import { mount } from 'enzyme';
import Pagination from '../pagination';
import { act } from 'react-dom/test-utils';
import { updateWrapper } from 'tests/utils';

describe('Pagination', () => {
  it('should render correctly', () => {
    const wrapper = mount(<Pagination />);
    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it('the specified page should be activated', async () => {
    const wrapper = mount(<Pagination total={10} initialPage={2} />);
    expect(wrapper.find('.active').text()).toEqual('2');
    await act(async () => {
      wrapper.setProps({ page: 10 });
    });
    await updateWrapper(wrapper, 200);
    expect(wrapper.find('.active').text()).toEqual('10');
  });

  it('should trigger change event', async () => {
    let current = 1;
    const handler = jest.fn().mockImplementation((val) => (current = val));
    const wrapper = mount(
      <Pagination total={10} initialPage={2} onChange={handler} />
    );

    await act(async () => {
      wrapper.setProps({ page: 10 });
    });
    await updateWrapper(wrapper, 200);
    expect(handler).toHaveBeenCalled();
    expect(current).toEqual(10);

    const btns = wrapper.find('button');
    btns.at(0).simulate('click');
    await updateWrapper(wrapper, 200);
    expect(current).toEqual(9);

    btns.at(btns.length - 1).simulate('click');
    btns.at(btns.length - 1).simulate('click');
    btns.at(btns.length - 1).simulate('click');
    btns.at(btns.length - 1).simulate('click');
    await updateWrapper(wrapper, 200);
    expect(current).toEqual(10);
    handler.mockRestore();
  });

  it('should trigger change event when ellipsis clicked', async () => {
    let current = 20;
    const handler = jest.fn().mockImplementation((val) => (current = val));
    const wrapper = mount(
      <Pagination total={20} initialPage={20} onChange={handler} />
    );
    const btn = wrapper.find('svg').at(0).parents('button');
    btn.at(0).simulate('click');
    await updateWrapper(wrapper, 200);
    expect(handler).toHaveBeenCalled();
    expect(current).toEqual(15);

    await act(async () => {
      wrapper.setProps({ page: 1 });
    });
    await updateWrapper(wrapper, 200);
    const lastBtn = wrapper.find('svg').at(0).parents('button');
    lastBtn.at(0).simulate('click');
    await updateWrapper(wrapper, 200);
    expect(current).toEqual(1 + 5);
  });

  it('another SVG should be displayed when the mouse is moved in', async () => {
    const wrapper = mount(<Pagination total={20} initialPage={20} />);
    const svg = wrapper.find('svg').at(0);
    const btn = svg.parents('button');

    const html = svg.html();
    btn.simulate('mouseEnter');
    await updateWrapper(wrapper);
    expect(html).not.toEqual(wrapper.find('svg').at(0).html());

    btn.simulate('mouseLeave');
    await updateWrapper(wrapper);
    expect(html).toEqual(wrapper.find('svg').at(0).html());
  });

  //   it('custom buttons should be display', () => {
  //     const wrapper = mount(
  //       <Pagination total={20}>
  //         <Pagination.Previous>custom-prev</Pagination.Previous>
  //         <Pagination.Next>custom-next</Pagination.Next>
  //       </Pagination>
  //     );
  //     const btns = wrapper.find('button');
  //     expect(btns.at(0).text()).toEqual('custom-prev');
  //     expect(btns.at(btns.length - 1).text()).toEqual('custom-next');
  //   });
});
