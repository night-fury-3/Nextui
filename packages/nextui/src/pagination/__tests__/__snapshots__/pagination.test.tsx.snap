// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render correctly 1`] = `"<nav class=\\"nextui-c-eKVPvf nextui-c-eKVPvf-jhIisp-color-default nextui-c-eKVPvf-dNflHU-size-md nextui-c-eKVPvf-iTJsWG-borderWeight-normal nextui-c-eKVPvf-flFLUl-onlyDots-false nextui-c-eKVPvf-bNGYLg-rounded-false nextui-c-eKVPvf-QbeIJ-noMargin-false nextui-pagination\\"><button class=\\"nextui-c-kVKzND nextui-c-huiNHE nextui-c-kVKzND-ccwhjb-disabled-true nextui-c-kVKzND-SWDEj-animated-true nextui-pagination-item nextui-pagination-item-animated nextui-pagination-item-disabled nextui-pagination-item-preserve-content\\" aria-label=\\"previous page button\\" tabindex=\\"-1\\"><span class=\\"nextui-c-fItrmj nextui-pagination-item-content\\"><svg role=\\"presentation\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"nextui-c-eKuzCY nextui-c-eKuzCY-xROYZ-isPrev-true nextui-pagination-icon\\"><path d=\\"M15.5 19l-7-7 7-7\\" stroke=\\"currentColor\\" stroke-width=\\"1.5\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"></path></svg></span></button><div aria-hidden=\\"true\\" class=\\"nextui-c-nTJos nextui-c-nTJos-befVxL-animated-true nextui-c-nTJos-igbOOjv-css nextui-pagination-highlight nextui-pagination-highlight--active\\"></div><button class=\\"nextui-c-kVKzND nextui-c-huiNHE nextui-c-kVKzND-dZWCtT-active-true nextui-c-kVKzND-SWDEj-animated-true nextui-pagination-item nextui-pagination-item-active nextui-pagination-item-animated\\" aria-label=\\"1 item active\\" tabindex=\\"0\\"><span class=\\"nextui-c-fItrmj nextui-pagination-item-content\\">1</span></button><button class=\\"nextui-c-kVKzND nextui-c-huiNHE nextui-c-kVKzND-ccwhjb-disabled-true nextui-c-kVKzND-SWDEj-animated-true nextui-pagination-item nextui-pagination-item-animated nextui-pagination-item-disabled nextui-pagination-item-preserve-content\\" aria-label=\\"next page button\\" tabindex=\\"-1\\"><span class=\\"nextui-c-fItrmj nextui-pagination-item-content\\"><svg role=\\"presentation\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"nextui-c-eKuzCY nextui-pagination-icon\\"><path d=\\"M15.5 19l-7-7 7-7\\" stroke=\\"currentColor\\" stroke-width=\\"1.5\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"></path></svg></span></button></nav>"`;
